### DISCUSSION TITLE
web editor stl export

### DISCUSSION THREAD

**Discussion by pca006132** [elalish/manifold#279] [Ideas]
It would be nice if we can directly export the model as a STL file, that way users can directly put it into a slicer. IIRC threejs already provides STL file exporter?

**Comment by elalish** [comment]
My issue with STL is it fundamentally can't transmit manifoldness (it has no connectivity info). I'd much prefer 3MF (partly because I helped develop it) as it solves this problem and I think most of the 3D printer slicers import it now.

**Reply by pca006132** [reply]
Any update on this? I wanted to mod my 3D printer so I did some design on the web editor but still have to convert it manually to other formats as slicers don't support glb.

**Reply by elalish** [reply]
I did some hunting around for an open-source JS 3MF library, but didn't find anything 😞. I suppose we could look at including assimp's 3MF support, though I'm not sure how much that would bloat our WASM. Of course it's just an XML format; it might not be that hard to write our own with an XML parser. Thoughts?

**Reply by hrgdavor** [reply]
I wrote a simple 3mf serializer for jscad
https://github.com/jscad/OpenJSCAD.org/tree/master/packages/io/3mf-serializer

I could extract it and tweak to work with manifold. It is basic but works for us when exporting jscad models

**Reply by elalish** [reply]
@hrgdavor That would be perfect, thank you!

**Reply by hrgdavor** [reply]
I will start with one on my git, as it also requires me to setup for myself a manifold sample ... I may run manifold web and try to plug there directly ...

is there a place inside manifold repo for such addons? ... after I create sth usable we can discuss where to place it.

**Reply by elalish** [reply]
Could it be a little stand-alone js file? Then it could live [here](https://github.com/elalish/manifold/tree/master/bindings/wasm/examples) and we could import it into `editor.js`.

I'm considering putting some file I/O support into `manifold.js`, so if this isn't too large, I may end up pulling it in there.

**Reply by elalish** [reply]
For a test sample, you may want to start with our simplest [example](https://github.com/elalish/manifold/blob/master/bindings/wasm/examples/three.html).

**Reply by hrgdavor** [reply]
ok, I will inject it as standalone file there

**Reply by pca006132** [reply]
Maybe we can just compile with assimp support for the web editor, and use MeshIO directly? assimp can export binary blob as well.

**Reply by elalish** [reply]
I have a lot of reservations about assimp, and I'm concerned about how much it would add to our WASM bundle size. I prefer `gltf-transform` on the web, especially with my new `EXT_manifold` extension. I'd like our file support to be as modular from our core library as possible. I still think a little 3MF JS library would be nice though.

**Reply by elalish** [reply]
@hrgdavor Any luck with integrating your 3MF library? Even if you don't have time to complete the integration, it would still be a huge help to make a PR to add your library to our repo so we could help get it over the finish line.

**Reply by hrgdavor** [reply]
@elalish hey, thanks for the reminder. I started to clean the code to be able to integrate, but then my schedule got soo crowded.

I misspoke when I told my code for 3mf was in jscad. My 3mf code is in my separate offline prototype, and I meant to get it into jscad, but main dev already had something ready at the time, so his variant is there. 

I looked at version that is in jscad, and is strongly bundled with jscad internals, but my scrappy version is more generic and also adds a thumbnail.

Hopefully I will be able to catch some breath in next few weeks, but I can not say it for certain. 

Maybe I should move the scrappy code and leave comments what it expects as input. This way somebody more familiar with manifold can take-over. At that point I would need to spend time understanding manifold internals to finish it (that part could be long or short, depends how fast I will understand it).

**Reply by elalish** [reply]
Great, thank you! Totally hear you on lack of time - just one of many reasons small PRs are better than big ones. Yes, I'd highly recommend you start with a PR to get the library in with comments. Then whoever is excited can work on integrating it, and that work may well come in several pieces too. I try to get PRs merged pretty quickly so that it's quick to do follow-on work without a bunch of conflicts.

**Reply by hrgdavor** [reply]
I started to clean it up a bit. https://github.com/hrgdavor/jscadui/tree/main/file-format/3mf-export

the minimum needed to produce a 3mf is relatively simple. It supports instancing, but unfortunately slicers like prusa do not respect it.

**Reply by elalish** [reply]
Excellent, thank you! When you say prusa does not respect instancing, what do you mean?

Do you plan to publish this to npm so we can take a dependency, or should we fork your code into our repo?

**Reply by hrgdavor** [reply]
>When you say prusa does not respect instancing, what do you mean?

I had a questionable use-case where I used multiple instances of a cube to produce a shape without doing boolean union. Prusa slicer does not then use those instances as a single model, but as separate entities and drops them all to the plate. I posted a bug report on prusa git. They were not willing to fix, then I told them "screw you" and closed the ticket ... not one of my finest moments :D.

details: https://github.com/prusa3d/PrusaSlicer/issues/7078

>Do you plan to publish this to npm so we can take a dependency, or should we fork your code into our repo?

The code was part of a quick prototype and messy. I finally started oragnizing the code so it is more easily reusable. I will publish it as npm package, and also put an example how to use `fflate` to create the actual zip file. `fflate` will intentionally be dev-dependecy so ppl using the 3mf lib can use their preferred zip library.

**Reply by elalish** [reply]
Ah, I see why they do that - well the good news is Manifold should take care of that problem nicely. Sounds like a good plan - any idea how long you think it'll take to get to a first npm release?

**Reply by hrgdavor** [reply]
I should have it tomorrow. I already have a npm acc and publishing my small libs regularly.

**Comment by pca006132** [comment]
That's fine either, as long as it can export something that can be imported by 3D printer slicers.

**Comment by Loosetooth** [comment]
I'd also be really interested in a (built-in or not) export mechanism that:
1. is compatible with slicers
2. keeps manifoldness

Having recently had a bit of a struggle with manifoldness problems, because of exporting in STL format, I think this would be an ideal addition to manifold.

@hrgdavor Could you share a link to your prototype 3mf code? Thanks.

**Reply by Loosetooth** [reply]
For the time being, when using three.js:

the [OBJExporter](https://threejs.org/docs/index.html?q=export#examples/en/exporters/OBJExporter) seems to work pretty well.

**Comment by hrgdavor** [comment]
Initial step of cleaning up the 3mf code I had in my prototype is done.
it is now published on npm https://www.npmjs.com/package/@jscadui/3mf-export
and code is at 
https://github.com/hrgdavor/jscadui/tree/main/file-format/3mf-export

it is now something that should work, but still has cleaning to do and documenting jsdoc types
